File: exercice4_1.py
import os

SIZECAR = 100

def main():
    tube1 = os.pipe()
    tube2 = os.pipe()

    pid_fils = os.fork()

    if pid_fils == -1:
        print("Erreur dans fork")
        return

    if pid_fils == 0:
        # ---------------- pipe1 ----------------
        caractereW = input("saisissez un mot\n")

        os.close(tube1[0])
        os.write(tube1[1], caractereW.encode())

        # ---------------- pipe2 ----------------
        os.close(tube2[1])

        resultat = os.read(tube2[0], 1024)
        resultat = int(resultat.decode())

        if resultat == 1:
            print("le mot existe dans le fichier")
        else:
            print("le mot n'existe pas dans le fichier")

    else:
        # ---------------- pipe1 ----------------
        os.close(tube1[1])

        caractereR = os.read(tube1[0], 1024).decode()

        compt = 0

        try:
            with open("fichier.txt", "r") as mot:
                for ligne in mot:
                    if ligne.strip() == caractereR:
                        compt = 1
                        break
        except FileNotFoundError:
            print("fichier.txt introuvable")
            os._exit(1)

        # ---------------- pipe2 ----------------
        os.close(tube2[0])
        os.write(tube2[1], str(compt).encode())

main()

File: exercice4_1.c
#include<stdio.h>
#include<stdlib.h>
#include<sys/wait.h>
#include<unistd.h>
#include<fcntl.h>
#include<string.h>

#define SIZECAR 100
int main()
{
    pid_t pid_fils;
    int tube1[2];
    int tube2[2];
    char caractereR[100],caractereW[100];

    //creation du pipe
    if(pipe(tube1)!=0 || pipe(tube2)!=0)
    {
        fprintf(stderr, "Erreur dans pipe\n");
        exit(1);
    }

    pid_fils = fork();

    if (pid_fils == -1)
    {
        fprintf(stderr, "Erreur dans fork\n");
        exit(1);
    }
    if(pid_fils == 0)
    {
        //----------------pipe1_-------------------
        printf("saisissez un mot\n");
        scanf("%s",caractereW);
        close(tube1[0]);
        write(tube1[1] , caractereW , SIZECAR);

        //---------------pipe_2--------------------

        int resultat;
        read(tube2[0],&resultat,sizeof(int));

        if(resultat == 1 )
        {
            printf("le mot existe dans le fichier\n");
        }
        else
        {
            printf("le mot n'existe pas dans le fichier\n");
        }

        close(tube2[1]);
    }
    else
    {
        //-------------------pipe_1----------------
        close(tube1[1]);
        read(tube1[0], caractereR , SIZECAR);
        FILE *mot;
        char ligne[100];
        int compt=0;

        mot=fopen("fichier.txt","r");//ouverture du fichier contenant le mot
        if(mot==NULL)
        {
            perror("fopen");
            exit(1);
        }

        while(fscanf(mot,"%s",ligne)!=EOF) //verification mot par mot
        {
            if(strcmp(ligne,caractereR)==0)//comparaison des chaines de caratcère
            {
                compt=1;
                break;
            }
        }

       fclose(mot);

       //---------------pipe_2--------------------
       write(tube2[1],&compt,sizeof(int));
        
        close(tube2[0]);

    }

    return(0);

}